### PR TITLE
Prevent field argument collision in diffTypeNodes

### PR DESCRIPTION
**Pull request by stevenschobert**
This MR addresses the bug outlined in #1100, where value types with similarly named field arguments result in a composition error.

I outlined some of the implementation details in my [issue comment](https://github.com/apollographql/federation/issues/1100#issuecomment-974504381), but the summary of the changes in this PR are:

* `diffTypeNodes` has been adjusted to consider arguments on a per-field basis
* `diffTypeNodes`'s return value has the following changes:
    ```diff
     {
       name: [],
       kind: [],
       fields: {},
    -  inputValues: {},
    +  fieldArgs: {},
       unionTypes: {},
       locations: [],
    -  args: {},
    +  directiveArgs: {}
     }
    ```
* All call sites to `diffTypeNodes` have been updated to handle the new return value's shape. In cases where the call site previously used the `inputValues` entry in the diff, a loop has been added to handle `fieldArgs` holding nested diffs per-field name.
* Added new unit test coverage for `diffTypeNodes`

### CODE REVIEW

**Comment by stevenschobert** [comment]
CLA is signed!

**Comment by trevor-scheer** [comment]
Thanks @stevenschobert! I've got a lot on my plate atm but I'll try to get to this as soon as I'm able. Might be tough with the holidays coming up - don't be afraid to ping me after the new year if I haven't given you a review.

**Comment by stevenschobert** [comment]
Hey @trevor-scheer! Just giving a friendly reminder about this PR, after the new year. Hope you had a wonderful break!

**Review by trevor-scheer — APPROVED** [review]
This is excellent, thanks for all the work you put into this @stevenschobert. Definitely appreciate all the new tests as well!

Would you mind adding a CHANGELOG entry to the federation package?

If you don't get to this in the next couple days, I'll take care of the rest and get this landed.

File: federation-js/src/composition/__tests__/utils.test.ts [hunk 1]
```diff
@@ -1,11 +1,581 @@
...
+          x(j: Int): String
+        }
+      `;
+
+      const defA = typeDefs.definitions[0] as TypeDefinitionNode;
+      const defB = typeDefs.definitions[1] as TypeDefinitionNode;
+
+      const result = diffTypeNodes(defA, defB);
+
+      expect(result).not.toEqual({
```
[the remark this answers is not in the document]
Comment (stevenschobert) [reply federation-js/src/composition/__tests__/utils.test.ts]: Great catch, thank you! Yeah, outdated test from my initial "figuring out the code".

I've corrected the test to check for the positive case.

**Comment by stevenschobert** [comment]
Thanks for the review! I've added a CHANGELOG entry, and corrected that negative test case.

**Comment by trevor-scheer** [comment]
LGTM! In case you're curious, I pushed up an empty commit to trigger CI - for some reason none of the circle checks ran on your most recent commit, or so it seemed. Looks good now. Thanks again! Been a pleasure working with you on this. 🙏

**Comment by trevor-scheer** [comment]
@stevenschobert this fix is now released in the `latest` federation packages 👍

**Pull request merged**